lib/generic/PoleEffect.hpp

```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include "Pole.hpp"
+
+class PoleEffect {
```

Comment:
This doesn't really feel like it needs to be a class. I would prefer just having two free functions, `SetPoleEffect` and `SetPoleLowerEffect`, and then `GetColor` can exist only in the .cpp file. It's perfectly fine to have functions not associated with a class in C++, and `PoleEffect::SetPoleEffect` is a pretty redundant name compared to just `SetPoleEffect`.

Reply:
I agree, I originally wrote it like that then shoved it in a class because everything else seemed to be in a class :grinning: I'll move it into free floating functions.

Reply:
My brain is melting with languages, does `GetColor` need to be `static` in that case or not?

Reply:
There are a few rules that you should follow in C++:

If it exists only in the source file, it should be in the unnamed namespace:

```
namespace {
int x = 0;
} // namespace
```

This applies to variables, functions, types, anything. The `static` keyword in practice means the same thing for functions and for variables defined at global scope. `namespace` is more general (it applies to classes as well), and it avoids confusion with the several other meanings of the `static` keyword.

If it exists only in the header file, it should be marked `inline`. Unfortunately, this is kind of inconsistent. Type definitions (classes and aliases) do not need to be marked inline (and in fact, they cannot). Function templates and constexpr functions are implicitly inline, and thus don't need the additional annotation. It looks like there's an issue with our compiler for doing this with variables, more on that later.

The remaining things should be declared `extern` in a header file, and then defined in the associated cpp file. This one is also inconsistent. Types do not need to be marked extern (and cannot be). Function declarations (`void f();`) are implicitly extern, and thus don't need the additional annotation.

So to summarize what annotations you actually need, anything that's just in a source file is inside the unnamed namespace, non-template, non-constexpr functions that are defined in the header are marked `inline`, and variables that are declared in the header and defined in the source file are marked `extern` in the header.

If you just declare a variable in the header file and it is not marked `inline` (which we apparently cannot do) and it is not in the unnamed namespace, then you are limited in how you can use it. If it is a compile-time constant (so it's marked `constexpr`, or it's an integer and it's marked `const` and it's initialized with a constant expression...), you can pass it by value to functions, but in any other case you can end up with linker errors and technically have undefined behavior. If you do put the variable in the unnamed namespace in a header, it means that every source file that includes that header has its own copy of the variable.

The `static` keyword has many meanings. One of them is the less flexible version of putting things in the unnamed namespace, which is what happens when you use `static` on something at global or namespace scope (I recommend never using this version of `static`). The second meaning is when you declare a member of a class `static`. This meaning is to have a global variable or normal free function, but you name it as `class_name::member_name` from outside the class, and you can make it `private` or `protected`.  The third meaning is when you declare a variable as `static` inside of a function. This means that the variable is initialized (in a thread-safe way) the first time the function is called, and the variable is destroyed after `main` finishes running.

Reply:
Man, there's a lot here. I think you are suggesting to put `GetColor` in the unnamed namespace so I'll go ahead and do that here.

Is there a linter of sorts that will enforce all these rules for us? Given some of the rules seem rather deterministic I don't see why I should think about them. :smile:

Reply:
I couldn't use the unnamed namespace wrapped around everything (including the external functions that use `GetColor`) so I went with static which felt appropriate. PTAL